### PR TITLE
Implement Codex exchange (pairs, order-book, ticker, trades)

### PR DESCRIPTION
**Pull request by EvgeniyMykhalichenko**
- What is the purpose of this Pull Request?
- What is the related issue for this Pull Request (if this PR fixes issue, prepend with "Fixes" or "Closes")?
- [x] I have added Specs
- [x] (If implementing Market Ticker) I have verified that the `volume` refers to BASE
- [x] (If implementing Market Ticker) I have verified that the `base` and `target` is assigned correctly
- [x] I have implemented the `trade_page_url` method that links to the exchange page with the `base` and `target` passed in. If not available, enter the root domain of the exchange website.
- [x] I have verified at least **ONE** ticker volume matches volume shown on the trading page (use script below)

```
client = Cryptoexchange::Client.new
pairs = client.pairs 'exchange_name'
tickers = pairs.map do |p| client.ticker p end
sorted_tickers = tickers.sort_by do |t| t.volume end.reverse
```

### CODE REVIEW

File: lib/cryptoexchange/exchanges/codex/market.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module Cryptoexchange::Exchanges
+  module Codex
+    class Market < Cryptoexchange::Models::Market
+      NAME = 'codex'
+      API_URL = 'https://api.codex.one'
+    end
```
Comment (tmlee) [inline lib/cryptoexchange/exchanges/codex/market.rb]: `trade_page_url` must be implemented even if there isn't one. You can follow the same method structure but return the url to the exchange root page.

**Comment by tmlee** [comment]
Specs are failing, you may want to delete the cassette and replay the query

File: lib/cryptoexchange/exchanges/codex/services/market.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        def adapt_all(output)
+          output['data'].map do |key, pair|
+            market_pair = Cryptoexchange::Models::MarketPair.new(
+                base:   pair['base_unit'],
+                target: pair['quote_unit'],
+                market: Codex::Market::NAME
+            )
+
+            adapt(market_pair, pair)
```
Comment (tmlee) [inline lib/cryptoexchange/exchanges/codex/services/market.rb]: `adapt(market_pair, pair)` inconsistent with the argument name below, causes confusion

File: lib/cryptoexchange/exchanges/codex/services/market.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+          ticker           = Cryptoexchange::Models::Ticker.new
+          ticker.base      = market_pair['base_unit']
+          ticker.target    = market_pair['quote_unit']
+          ticker.market    = Codex::Market::NAME
+          ticker.last      = NumericHelper.to_d(market_pair['last'])
+          ticker.high      = NumericHelper.to_d(market_pair['high'])
+          ticker.low       = NumericHelper.to_d(market_pair['low'])
+          ticker.volume    = NumericHelper.to_d(market_pair['volume'])
+          ticker.timestamp = nil
+          ticker.payload   = output
```
Comment (tmlee) [inline lib/cryptoexchange/exchanges/codex/services/market.rb]: in doing so, the payload is going to store market_pair instead of the API response

Comment (EvgeniyMykhalichenko) [reply lib/cryptoexchange/exchanges/codex/services/market.rb]: +

File: lib/cryptoexchange/exchanges/codex/services/market.rb (hunk 2 above)
Comment (EvgeniyMykhalichenko) [reply lib/cryptoexchange/exchanges/codex/services/market.rb]: done

File: lib/cryptoexchange/exchanges/codex/market.rb (hunk 1 above)
Comment (EvgeniyMykhalichenko) [reply lib/cryptoexchange/exchanges/codex/market.rb]: done

**Comment by wongy91** [comment]
LGTM

**Pull request merged**